### PR TITLE
Bug 1222459 - Invalidate top site cache after removing an item

### PR DESCRIPTION
**Pull request by sleroux**

### CODE REVIEW

File: Client/Frontend/Home/TopSitesPanel.swift [hunk 1]
```diff
@@ -133,13 +133,19 @@ class TopSitesPanel: UIViewController {
...
+                self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount).uponQueue(dispatch_get_main_queue(), block: { result in
+                    self.deleteOrUpdateSites(result, indexPath: indexPath)
+                })
+            }
         }
     }
 
+    private func invalidateTopSites() -> Success {
+        profile.history.setTopSitesNeedsInvalidation()
+        return profile.history.invalidateTopSitesIfNeeded() >>> { succeed() }
```
Comment (rnewman) [inline Client/Frontend/Home/TopSitesPanel.swift]: Nit: `>>> { succeed() }` :arrow_right: `>>> succeed`.

Substantive: can we avoid the two pref writes by defining:

```
… SQLiteHistory {
    public func refreshTopSitesCache() -> Success {
        let cacheSize = Int(prefs.intForKey(PrefsKeys.KeyTopSitesCacheSize) ?? 0)
        return self.clearTopSitesCache()
            >>> { self.updateTopSitesCacheWithLimit(cacheSize) }
    }
```

and using that both here and inside `invalidateTopSitesIfNeeded`?

Comment (rnewman) [inline Client/Frontend/Home/TopSitesPanel.swift]: … or similar. 

Also, I think `invalidateTopSitesIfNeeded` is misnamed. It should really be `updateTopSitesCacheIfInvalidated`, right?

File: Client/Frontend/Home/TopSitesPanel.swift [hunk 2]
```diff
@@ -133,13 +133,19 @@ class TopSitesPanel: UIViewController {
 
     private func deleteHistoryTileForSite(site: Site, atIndexPath indexPath: NSIndexPath) {
         profile.history.removeSiteFromTopSites(site) >>== {
-            self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount).uponQueue(dispatch_get_main_queue(), block: { result in
-                self.updateDataSourceWithSites(result)
-                self.deleteOrUpdateSites(result, indexPath: indexPath)
-            })
+            self.invalidateTopSites() >>== {
```
Comment (rnewman) [inline Client/Frontend/Home/TopSitesPanel.swift]: A better name is `updateTopSitesCache`, and stacking like this makes it clearer:

```
    func reloadThumbnails() {
        self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
                            .uponQueue(dispatch_get_main_queue()) {
                                self.deleteOrUpdateSites(result, indexPath: indexPath)
                            }
    }

    profile.history.removeSiteFromTopSites(site)
>>> self.updateTopSitesCache
>>> reloadThumbnails
```

File: Client/Frontend/Home/TopSitesPanel.swift (hunk 1 above)
Comment (sleroux) [inline Client/Frontend/Home/TopSitesPanel.swift]: Ya I suppose we could name it that. I was following the way UIKit names it's invalidation routines like layoutIfNeeded and setNeedsLayout.

Comment (rnewman) [inline Client/Frontend/Home/TopSitesPanel.swift]: It's less the `Needed` part and more the `invalidated` part that concerns me — that method should be called *after* something has already set the invalidation-needed pref, and it's responsible for dropping the old cache and repopulating it.

Comment (sleroux) [inline Client/Frontend/Home/TopSitesPanel.swift]: Makes sense - I'll update

**Comment by sleroux** [comment]
@rnewman Updated PR with deferred fixes, API clean up, and fix for https://bugzilla.mozilla.org/show_bug.cgi?id=1223538.

File: Client/Frontend/Home/TopSitesPanel.swift [hunk 3]
```diff
@@ -132,12 +132,17 @@ class TopSitesPanel: UIViewController {
...
 
     private func deleteHistoryTileForSite(site: Site, atIndexPath indexPath: NSIndexPath) {
-        profile.history.removeSiteFromTopSites(site) >>== {
-            self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount).uponQueue(dispatch_get_main_queue(), block: { result in
-                self.updateDataSourceWithSites(result)
-                self.deleteOrUpdateSites(result, indexPath: indexPath)
-            })
+        func reloadThumbnails() -> Success {
+            return self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
+                .bindQueue(dispatch_get_main_queue()) { result in
```
Comment (rnewman) [inline Client/Frontend/Home/TopSitesPanel.swift]: We don't need to return a value here — it never escapes `deleteHistoryTileForSite`, and it's not used subsequently within, so you can just use `uponQueue` and not bother with any of the `return`s.

File: Client/Frontend/Home/TopSitesPanel.swift [hunk 4]
```diff
@@ -169,6 +174,11 @@ class TopSitesPanel: UIViewController {
         // now update the data source with the new data
         self.updateDataSourceWithSites(result)
 
+        // Exit early if the query failed in some way.
+        guard result.isSuccess else {
```
Comment (rnewman) [inline Client/Frontend/Home/TopSitesPanel.swift]: You can actually move this before the `self.updateDataSourceWithSites` call, too.

File: Storage/SQL/SQLiteHistory.swift [hunk 5]
```diff
@@ -290,6 +290,12 @@ extension SQLiteHistory: BrowserHistory {
         }
     }
 
+    public func refreshTopSitesCache() -> Success {
```
Comment (rnewman) [inline Storage/SQL/SQLiteHistory.swift]: Use this in `updateTopSitesCacheIfInvalidated`.

**Comment by sleroux** [comment]
@rnewman Updated with your termination commit and fixes.

**Pull request merged**